### ISSUE TITLE
The `--quiet` option doesn't suppress progress messages

### ISSUE THREAD

**Issue by akaihola** [nexB/scancode-toolkit#417]
The `scancode --help` instructions advise to use the `--quiet` option to skip printing progress messages:

    $ ./scancode --help
    ...
      --quiet                  Do not print progress messages.
    ...

However, that option has no effect. I still get the progress messages printed on stdout before the JSON output. This prevents me from running `scancode` as a subprocess and capturing its output as JSON:

    $ ./scancode --quiet -l /tmp/LICENSE
    Scanning files for: licenses with 1 process(es)...
    Building license detection index...Done.
    Scanning files...
    Scanning done.
    Scan statistics: 1 files scanned in 3s.
    Scan options:    licenses with 1 process(es).
    Scanning speed:  0.57 files per sec.
    Scanning time:   1s.
    Indexing time:   1s.
    Saving results.
    {... result JSON ...}

**Comment by pombredanne** [comment]
Thank you for reporting this! This is closely related to #388 from @sschuberth that is fixed but still needs work to implement the following scheme:

1. if no output file is specified, send the scan results to `stdout` 
2. with no option: messages, progress bar and summary sent to `stderr`. 
3. with the option `--quiet` no messages, no progress bar, no summary 
4. with the option `--verbose` messages, per-file log and  summary sent to `stderr`

The key would be to always send messages to `stderr` and the scan output to `stdout`
@akaihola would this work for you alright?

**Comment by akaihola** [comment]
@pombredanne, yes, that would be perfect. 

Actually, before noticing the `--quiet` option, I was expecting to be able to read the JSON output from stdout and discard messages in stderr.

I'm currently working around this by doing the reverse: specify `/dev/stderr` as the output file, read JSON output from stderr and discard messages in stdout.

**Comment by pombredanne** [comment]
@akaihola No nee to work around... I am working on a  proper fix and will push it in a few

**Comment by pombredanne** [comment]
@akaihola @sschuberth  The branch `388-417-consistent-stderr-stdout-outputs` contains a proper implementation of the proposed update to outputs redirection for here and #388.
Feedback welcomed.

**Comment by pombredanne** [comment]
@akaihola Just curious you said:
> This prevents me from running scancode as a subprocess and capturing its output as JSON

what is your use case there? I am always interested by new ways to use scancode!

**Comment by akaihola** [comment]
@pombredanne Our use case is a Python script which
- accepts a list of Python package names (and optional versions) and/or paths to Python package source distribution tarballs/zipfiles
- downloads named packages from a package index server
- from existing and downloaded packages, extracts files whose names look like license files into a temporary directory
- reads license information from package metadata using `pkginfo`
- also identifies license(s) from extracted license files using [Ninka](http://ninka.turingmachine.org/) and/or Scancode
- outputs a JSON license summary of the packages

I'm planning to suggest to my employer that we open source the script.

**Comment by pombredanne** [comment]
@akaihola Thanks for testing so quickly!

You wrote:
> however, the progress bar is still written to stdout:

I will have to digg if this is an option with the click progressbar at all... if not we cqan consider using https://pypi.python.org/pypi/tqdm/ instead if it can echo to stderr (or patch and push a PR to click)

You use case is very interesting. A few  notes:
1. afaik, there should not be a case where Ninka detects something that is not detected by ScanCode. If there is one, I would consider this as a bug to be fixed ASAP.

2. There is code in this branch https://github.com/nexB/scancode-toolkit/tree/253-python-packages that may not do exactly what you are looking for but is working towards it... And upcoming code to actually also fetch Pypi (and other package repos)

3. If you only look at files that look like a license, you may be missing out a few things IMHO (and FWIW there is simple code in scancode to do that here: https://github.com/nexB/scancode-toolkit/blob/538648e7bec139ee60e0ddf76481c0cda40eab84/src/licensedcode/legal.py but that has not yet been plugged in a scan proper

**Comment by pombredanne** [comment]
Actually I think I could simply override `render_progress` in https://github.com/pallets/click/blob/30fc626263db82e9f339ca8810d207ab089d5575/click/_termui_impl.py#L193 to ensure that the progress bar goes to `stderr` alright.

**Comment by pombredanne** [comment]
And the solution to send the progressbar to stderr was actually quite simple: Click's bar have a `file` argument that defaults to stdout and I just passed it sys.stderr

**Comment by akaihola** [comment]
@pombredanne, now the branch is perfect!

After running a good number of packages through scancode, I concluded that I indeed can ditch Ninka and just rely on scancode.

I am encountering suspicious results with scancode for a few packages which have either heavily modified licenses or have come up with their own license. Some examples:
- pyproj 1.9.5.1
- Matplotlib 1.5.3 (licenses of some of the included components, e.g. stix, yorick, pnpoly, bakoma, qhull, agg24)
- Pillow 3.4.2

Is there a better place for reporting these? As separate issues maybe?

Thanks for the 253-python-packages tip, I'll try it out. Any pointers to usage instructions yet?

Only looking at the license text files speeds up scanning and is sufficient for us at this point – we'll submit everything for manual review anyway.

**Comment by sschuberth** [comment]
> Is there a better place for reporting these? As separate issues maybe?

Yes, please report each of these as a separate issue, with as much info to reproduce as possible (like the ScanCode version and command line arguments used, the exact Git SHA1 of the source code being scanned etc.)

**Comment by pombredanne** [comment]
@akaihola Great!
I am closing this then.

You wrote:

> I am encountering suspicious results with scancode for a few packages which have either heavily modified licenses or have come up with their own license.  [...] Is there a better place for reporting these? As separate issues maybe?

I entered #419 for this. You can either add to that new ticket or create new ones for each issue you may encounter. I will push separate commit for each problem anyway. Adding a new license detection rule is rather easy (hopefully) and documented here: https://github.com/nexB/scancode-toolkit#how-to-add-a-new-license-for-detection and here: https://github.com/nexB/scancode-toolkit/wiki/How-to:-Add-new-license-rules-for-enhanced-detection and here: https://github.com/nexB/scancode-toolkit/wiki/FAQ

> Thanks for the 253-python-packages tip, I'll try it out. Any pointers to usage instructions yet?

Given a set of installed packages (or extracted with `extractcode`) you just need to use the the `-p` or `--package` option to scan for and detect packages. I am planning to merge this branch in develop anyway today, as I have a bunch of serious commits in the package area pending and this would make this branch go stale and diverging too much. It is far from what I want for Python detection, but a good start anyway.

> Only looking at the license text files speeds up scanning and is sufficient for us at this point – we'll submit everything for manual review anyway.

Fair enough, but I usually prefer looking at everything... That said: scancode can run in parallel on multiple processes for serious speedups (assuming you use the develop branch) using the the `-n x` or `--processes x` command line option where `x` is the number of processes to spawn.
